src/Umbraco.Web/Editors/MacrosController.cs

```diff
@@ -0,0 +1,320 @@
+    {
+        /// <summary>
+        /// Creates a new macro
+        /// </summary>
+        /// <param name="name">
+        /// The name.
+        /// </param>
+        /// <returns>
+        /// The <see cref="HttpResponseMessage"/>.
+        /// </returns>
+        [HttpPost]
+        public HttpResponseMessage Create(string name)
+        {
+            if (string.IsNullOrWhiteSpace(name))
+            {
+                return this.ReturnErrorResponse("Name can not be empty");
+            }
+
+            var alias = name.ToSafeAlias();
+
+            if (this.Services.MacroService.GetByAlias(alias) != null)
+            {
+                return this.ReturnErrorResponse("Macro with this alias already exists");
+            }
+
+            try
+            {
+                var macro = new Macro
+                {
+                    Alias = alias,
+                    Name = name,
+                    MacroSource = string.Empty,
+                    MacroType = MacroTypes.PartialView
+                };
+
+                this.Services.MacroService.Save(macro, this.Security.CurrentUser.Id);
+
+                return this.Request.CreateResponse(HttpStatusCode.OK, macro.Id);
+            }
+            catch (Exception exception)
+            {
+                return this.ReturnErrorResponse("Error creating macro", true, exception);
+            }
+        }
+
+        [HttpGet]
+        public HttpResponseMessage GetById(int id)
+        {
+            var macro = this.Services.MacroService.GetById(id);
+
+            if (macro == null)
+            {
+                return this.ReturnErrorResponse($"Macro with id {id} does not exist");
+            }
+
+            var macroDisplay = new MacroDisplay
+            {
+                Alias = macro.Alias,
+                Id = macro.Id,
+                Key = macro.Key,
+                Name = macro.Name,
+                CacheByPage = macro.CacheByPage,
+                CacheByUser = macro.CacheByMember,
+                CachePeriod = macro.CacheDuration,
+                View = macro.MacroSource,
+                RenderInEditor = !macro.DontRender,
+                UseInEditor = macro.UseInEditor,
+                Path = $"-1,{macro.Id}"
+            };
+
+            var parameters = new List<MacroParameterDisplay>();
+
+            foreach (var param in macro.Properties.Values.OrderBy(x => x.SortOrder))
+            {
+                parameters.Add(new MacroParameterDisplay
+                {
+                    Editor = param.EditorAlias,
+                    Key = param.Alias,
+                    Label = param.Name,
+                    Id = param.Id
+                });
+            }
+
+            macroDisplay.Parameters = parameters;
+
+            return this.Request.CreateResponse(HttpStatusCode.OK, macroDisplay);
+        }
+
+        [HttpPost]
+        public HttpResponseMessage Save(MacroDisplay macroDisplay)
```

Comment:
The reason why we separate model names by "Display" and "Save" is because in many cases we aren't posting the same model that is received because we don't want to 'over post' data. In this case, it's probably the same model, in which case the name could just be Macro, MacroModel, MacroEditingModel. Not a big deal but just wanted to mention why the names are as they are. I'll leave for now since maybe in the future we'll end up with a MacroSave to post the data.